File: 2018/go/problem01.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func parseInput() []int {
	file, err := os.Open("../inputs/01-input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var numbers []int
	for scanner.Scan() {
		number, err := strconv.Atoi(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
		numbers = append(numbers, number)
	}
	return numbers
}

func part1() int {
	numbers := parseInput()
	result := 0
	for _, v := range numbers {
		result += v
	}
	return result
}

func part2() int {
	numbers := parseInput()
	result := 0
	results := map[int]bool{0: true}
	for i := 0; ; i = (i + 1) % len(numbers) {
		result += numbers[i]
		_, has_result := results[result]
		if has_result {
			return result
		}
		results[result] = true
	}
}

func main() {
	fmt.Println(part1())
	fmt.Println(part2())
}

File: 2018/rust/src/problem01.rs
use std::collections::HashSet;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

fn parse_input() -> Result<Vec<i32>, Box<Error>> {
    let buf_reader = BufReader::new(File::open("../inputs/01-input.txt")?);
    // Commented this out because of ugly double unwrap
    // let numbers = buf_reader
    //     .lines()
    //     .map(|line| line.unwrap().parse().unwrap())
    //     .collect();
    let mut numbers = vec![];
    for line in buf_reader.lines() {
        numbers.push(line?.parse()?);
    }
    Ok(numbers)
}

pub fn part1() -> Result<i32, Box<Error>> {
    let result = parse_input()?.iter().sum();
    Ok(result)
}

pub fn part2() -> Result<i32, Box<Error>> {
    let mut result = 0;
    let mut results = HashSet::new();
    results.insert(result);
    let numbers = parse_input()?;
    for n in numbers.iter().cycle() {
        result += n;
        if results.contains(&result) {
            break;
        }
        results.insert(result);
    }
    Ok(result)
}
